R/calc.dtsurvey.R

```diff
@@ -0,0 +1,364 @@
+#' @noRd
+compute = function(DT, x, by = NULL, metrics, ci_method = 'mean', level = .95, time_var, time_format, per = 1, window = FALSE){
+
+  sv = attr(DT, 'sdes')
+  st = attr(DT, 'stype')
+
+  #For each metric, define a function to compute it-- or ignore it if not called for.
+  xisfactor = is.factor(DT[,x,env = list(x=x)])
+  #construct the query
+  if(any(c('mean', 'rate') %in% metrics)){
+    mean_fun = data.table::substitute2(list(dtsurvey::smean(x,
+                                                  na.rm = T,
+                                                  var_type = c('se', 'ci'),
+                                                  ci_method = cim,
+                                                  level = l,
+                                                  ids = `_id`,
+                                                  sv = ..sv,
+                                                  st = ..st)),
+                                       list(x = x,
+                                            l = level,
+                                            cim = I(ci_method)))
+  }else{
+    mean_fun = NULL
+  }
+
+  if('total' %in% metrics){
+    total_fun = data.table::substitute2(list(dtsurvey::stotal(x,
+                                                    na.rm = T,
+                                                    var_type = c('se', 'ci'),
+                                                    ci_method = 'total',
+                                                    level = l,
+                                                    ids = `_id`,
+                                                    sv = ..sv,
+                                                    st = ..st)),
+                                        list(x = x,
+                                             l = level))
+  }else{
+    total_fun = NULL
+  }
+
+  #numerator
+  if('numerator' %in% metrics){
+    num_fun = data.table::substitute2(sum(x,na.rm = T), list(x = x))
+  }else{
+    num_fun = NULL
+  }
+  #denominator
+  if('denominator' %in% metrics){
+    denom_fun = data.table::substitute2(sum(!is.na(x)), list(x = x))
+  }else{
+    denom_fun = NULL
+  }
+
+  #obs (see below)
+
+  #median
+  if('median' %in% metrics){
+    if(xisfactor){
+      med_fun = NULL
+      warning('Ignoring a request to calculate the median on a factor')
+    }else{
+      med_fun = data.table::substitute2(median(x, na.rm = T) * 1.0, list(x=x))
+    }
+  }else{
+    med_fun = NULL
+  }
+
+  #time var
+  if(!is.null(time_var)){
+    time_fun = data.table::substitute2(time_format(time_var[!is.na((x))]), list(time_var = time_var, x=x))
+
+    #if we're in a window, don't "by" by time var. Instead, let time_format handle things
+    #if(!window) by = c(by, time_var) #add time_var to by is specified
+  }else{
+    time_fun = NULL
+  }
+
+  #unique.time
+  if('unique.time' %in% metrics){
+    ut_fun = data.table::substitute2(length(unique( (tv)[!is.na(x)] )), list(tv = time_var, x =x ))
```

Comment:
where does tv come from? from line 86 above? Is there value in explicitly adding it as an argument?